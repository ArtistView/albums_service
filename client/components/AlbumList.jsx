import React from 'react';
import ReactDom from 'react-dom';
import Album from './Album.jsx';

const AlbumList = (props) => {
  console.log('hello world');
  return <div>Hello World</div>;
};

export default AlbumList;