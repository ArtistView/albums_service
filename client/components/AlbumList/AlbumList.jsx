import React from 'react';
import $ from 'jquery';
import Promise from 'bluebird';
import styled, { css } from 'styled-components';
import {AiOutlineDown, AiOutlineUp} from 'react-icons/ai';
import Album from '../Album/Album.jsx';

const AlbumListWrapper = styled.div`
  padding-left: 25px;
  background-color: #151515;
`;

const AlbumListTitle = styled.div`
  font-size: 28px;
  padding-top: 65px;
  margin-bottom: 20px;
  margin-left: 10px;
  margin-right: 10px;
`;

const AlbumListShowMoreLess = styled.div`
  letter-spacing: 2px;
  clear: both;
  text-align: center;
  margin-top: 20px;
  font-size: 11px;
  font-weight: normal;
  position: relative;
`;

class AlbumList extends React.Component {
  // this component accepts a list of albums as props along with a playing function and currPlaying variable that are just needed to pass down to the album component, so it can modify the global state of which album is playing
  // if the type is Appears on, the albums prop is already all the featured albums, otherwise it is all albums that are primarily by that artist regardless of type (thus we will filter by type in componentDidMount)
  constructor(props) {
    super(props);
    this.state = {
      type: this.props.type, // stores the type of album this list is showing
      albums: [], // stores the list of albums that match the type
      showMore: false, // boolean telling us whether or not we should show more than 12 albums
      showMoreText: '', // stores the text to be rendered within the show more/less button
      showMoreArrow: '', // stores the icon to be rendered with the show more/less button
      showAtAll: false, // boolean telling us whether to show the component at all, should be false only if there are no albums of this particular type
      currShowing: [], // stores the list of albums to show, could be all the albums or just the first 12 depending on showMore
      numToShow: 12, // stores the number of albums that should be shown when show more is not pressed, should change when the window shrinks
      // TODO: update numToShow when the window shrinks and resize the albums to fit the window
    };
    this.showMoreLess = this.showMoreLess.bind(this);
    this.updateNumShowing = this.updateNumShowing.bind(this);
    this.populateAlbums = this.populateAlbums.bind(this);
  }

  async componentDidMount() {
    // upon loading we loop through all the albums from the artist and add only the ones that match our type to the array, which is then set to be albums in state
    if ($(window).width() < 1200) {
      await this.updateNumShowing();
      // if the screen is smaller, update the number and wait for it to finish before moving on
    }
    this.populateAlbums();
    // populate the state with the correct albums
  }
  // need to write tests

  populateAlbums() {
    // populates the state with albums that match the type
    let albumsOfType = [];
    if (this.state.type === 'Appears On') {
      albumsOfType = this.props.albums;
    } else if (this.state.type === 'Albums') {
      for (var album of this.props.albums) {
        if (album.type === 'album') {
          albumsOfType.push(album);
        }
      }
    } else if (this.state.type === 'Singles and EPs') {
      for (var album of this.props.albums) {
        if (album.type === 'single' || album.type === 'EP') {
          albumsOfType.push(album);
        }
      }
    } else {
      for (var album of this.props.albums) {
        if (album.type === 'compilation') {
          albumsOfType.push(album);
        }
      }
    }
    // if there are any albums of this type, show them and set the list equal to albums
    if (albumsOfType.length > 0) {
      this.setState({
        albums: albumsOfType,
        showAtAll: true,
        currShowing: albumsOfType,
      });
    }
    // if there are more than 12, make sure the show more button appears and we start by only showing the first 12 albums
    if (albumsOfType.length > this.state.numToShow) {
      this.setState({
        showMoreText: 'SHOW MORE',
        currShowing: albumsOfType.slice(0, this.state.numToShow),
        showMoreArrow: <AiOutlineDown id="album-down-arrow" />,
      });
    }
    window.addEventListener('resize', this.updateNumShowing);
    // adds an event listener that updates the number of albums showing if the window resizes
  }

  updateNumShowing() {
    // set number to be the number of elements shown in the first two rows
    let number = 12;
    const width = $(window).width();
    if (width > 1040 && width < 1200) {
      number = 8;
    } else if (width < 1040 && width > 850) {
      number = 3;
    } else if (width < 850) {
      number = 2;
    } else {
      number = 12;
    }
    if (this.state.showMore === true) {
      this.setState({
        numToShow: number,
        currShowing: this.state.albums,
      });
    } else {
      this.setState({
        numToShow: number,
        currShowing: this.state.albums.slice(0, number),
      });
    }

  }

  showMoreLess(event) {
    if (this.state.showMore === false) {
      // if it's currently showing twelve (or less if the screen is smaller) make it show all of the albums and change the text
      this.setState({
        showMore: true,
        showMoreText: 'SHOW LESS',
        currShowing: this.state.albums,
        showMoreArrow: <AiOutlineUp id="album-up-arrow" />,
      });
    } else {
      // if it's currently showing all the albums make it only show 12 (numToShow) and change the text
      this.setState({
        showMore: false,
        showMoreText: 'SHOW MORE',
        currShowing: this.state.albums.slice(0, this.state.numToShow),
        showMoreArrow: <AiOutlineDown id="album-down-arrow" />,
      });
    }
  }

  render() {
    // the albums list component renders a title, which is the type, a list of album components and a show more/less button with clickable functionality
    if (this.state.showAtAll) {
      return(
        <AlbumListWrapper className="album-list">
          <AlbumListTitle className="album-list-title">{this.state.type}</AlbumListTitle>
          <div className="album-list-list">
            {this.state.albums.map((album) => {
              // want to rerun this map function when currShowing changes
              let showing = false;
              if (this.state.currShowing.includes(album)) {
                showing = true;
              }
              return (<Album album={album} key={album._id} id={album._id} startPlaying={this.props.playing} currPlaying={this.props.currPlaying} show={showing} />);
            })
            }
          </div>
          <AlbumListShowMoreLess className="album-list-show-more-less" onClick={this.showMoreLess}>
            {this.state.showMoreText}
            {this.state.showMoreArrow}
          </AlbumListShowMoreLess>
        </AlbumListWrapper>
      );
    }
    return <div id="no-albums-of-this-type" />;
  };
};

export default AlbumList;
