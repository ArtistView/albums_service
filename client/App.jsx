import React from 'react';
import ReactDom from 'react-dom';
import styled, { css } from 'styled-components';
import AlbumList from './components/AlbumList/AlbumList.jsx';

const MainAlbumsWrapper = styled.div`
  color: white;
  background-color: #151515;
  font-family: proxima-nova, sans-serif;
  font-style: normal;
  font-weight: bold;
  margin-bottom: 90px;
`;
// margin-left: 232px;

class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      artist: '5eebdbf1bf2d490c13ff868d', // this is hardcoded but I should pull this state down from the url
      // artist: '5eebe7b2c451ef0e3f4c04f4', // another artist to use but this is has fewer albums
      albums: [], // set this initally to an empty array, represents the albums written by the artist
      features: [], // set to an empty array initially, represents the ablums that the artist appears on
      isLoaded: false, // used for conditional rendering to make sure the fetch has completed before rendering
      playing: '', // need to know which album is playing so all other albums can be set to stop playing
    };
    this.getAlbums = this.getAlbums.bind(this);
    this.getFeatures = this.getFeatures.bind(this);
    this.playing = this.playing.bind(this);
    // this.updateAudio = this.updateAudio.bind(this);
  }

  componentDidMount() {
    this.getAlbums(this.state.artist);
    this.getFeatures(this.state.artist);
    // upon loading the app, we grab the albums and features of this artist and store them in the state
  }

  getAlbums(artist) {
    const url = `http://localhost:3273/albums/${artist}`;
    // calling a get request to pull down all the albums in the db that matches the artist
    // this only pulls down the albums that the artist is the primary artist for, not appears on
    fetch(url)
      .then((response) => response.json())
      .then((data) => {
        this.setState({
          albums: data,
        });
      })
      .catch((err) => {
        console.error(err);
      });
  }

  getFeatures(artist) {
    const url = `http://localhost:3273/albums/features/${artist}`;
    // calling a get request to pull down all the albums in the db whose featuredArtists array contains the artist
    // this only pulls down the albums that the artist appears on
    fetch(url)
      .then((response) => response.json())
      .then((data) => {
        this.setState({
          features: data,
          isLoaded: true,
        });
      })
      .catch((err) => {
        console.error(err);
      });
  }

  playing(album) {
    // this method will be passed as a prop to the lower components so they can change this global state
    // need to track which album is playing and be able to change it within the album component so no two albums can play at the same time
    this.setState({
      playing: album,
    });
  }

  render() {
    if (this.state.isLoaded) {
      return (
        <MainAlbumsWrapper id="main-albums-wrapper">
          <AlbumList type="Albums" albums={this.state.albums} playing={this.playing} currPlaying={this.state.playing} />
          <AlbumList type="Singles and EPs" albums={this.state.albums} playing={this.playing} currPlaying={this.state.playing} />
          <AlbumList type="Compilations" albums={this.state.albums} playing={this.playing} currPlaying={this.state.playing} />
          <AlbumList type="Appears On" albums={this.state.features} playing={this.playing} currPlaying={this.state.playing} />
        </MainAlbumsWrapper>
      );
    }
    return (
      <div />
    );
  }
}

export default App;
