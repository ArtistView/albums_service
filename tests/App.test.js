/**
 * @jest-environment jsdom
*/
// import 'jsdom-global/register';
import React from 'react';
import ReactDom from 'react-dom';
import '../test-setup.js';
import { shallow, mount } from 'enzyme';
import App from '../client/App.jsx';
import AlbumList from '../client/components/AlbumList/AlbumList.jsx';
import {fetch} from 'whatwg-fetch';

describe('App', () => {
  const wrapper = shallow(<App />);
  it('renders correctly in \'debug mode\' ', () => {
    const component = shallow(<App debug />);
    expect(component).toMatchSnapshot();
  });
  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('has state', () => {
    expect(wrapper.state('artist')).toBe('5eebdbf1bf2d490c13ff868d');
    expect(wrapper.state('playing')).toBe('');
  });
});
