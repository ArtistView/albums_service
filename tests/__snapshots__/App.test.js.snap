// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `ShallowWrapper {}`;

exports[`App renders correctly in 'debug mode'  1`] = `ShallowWrapper {}`;
